core/primitives/src/shard_layout.rs

```diff
@@ -620,6 +620,45 @@ impl ShardLayout {
         Ok(parent_shard_id)
     }
 
+    /// Derive new shard layout from an existing one
+    pub fn derive_shard_layout(
+        base_layout: &ShardLayout,
+        new_boundary_account: AccountId,
+    ) -> ShardLayout {
+        let mut boundary_accounts = base_layout.boundary_accounts().clone();
+        let mut shard_ids = base_layout.shard_ids().collect::<Vec<_>>();
+        let mut shards_split_map = shard_ids
+            .iter()
+            .map(|id| (*id, vec![*id]))
+            .collect::<BTreeMap<ShardId, Vec<ShardId>>>();
+
+        debug_assert!(
+            boundary_accounts.iter().position(|acc| acc == &new_boundary_account).is_none(),
+            "duplicated boundary account",
+        );
+
+        // boundary accounts should be sorted such that the index points to the shard to be split
+        boundary_accounts.push(new_boundary_account.clone());
+        boundary_accounts.sort();
+        let new_boundary_account_index = boundary_accounts
+            .iter()
+            .position(|acc| acc == &new_boundary_account)
+            .expect("account should be guaranteed to exist at this point");
+
+        // new shard ids start from the current max
+        let max_shard_id =
+            *shard_ids.iter().max().expect("there should always be at least one shard");
+        let new_shards = vec![max_shard_id + 1, max_shard_id + 2];
+        let parent_shard_id = shard_ids
+            .splice(new_boundary_account_index..new_boundary_account_index + 1, new_shards.clone())
+            .collect::<Vec<_>>();
```

Comment:
mini nit: Personally I like the `collect_vec` method more - you may need to add dependency on itertools.

Reply:
I'm more of a minimal-dependency guy :))